% Generated by roxygen2 (4.1.1): do not edit by hand
% Please edit documentation in R/breakpointer.R
\name{breakpointer}
\alias{breakpointer}
\title{Wrapper function for the breakpointR package}
\usage{
breakpointer(inputfolder, outputfolder, configfile = NULL, numCPU = 1,
  reuse.existing.files = TRUE, windowsize = 100, scaleWindowSize = TRUE,
  pairedEndReads = FALSE, chromosomes = NULL,
  remove.duplicate.reads = TRUE, min.mapq = 10, trim = 10,
  peakTh = 0.33, zlim = 3.291, bg = 0.02, minReads = 10,
  WC.cutoff = 0.9)
}
\arguments{
\item{inputfolder}{Folder with either BAM files.}

\item{outputfolder}{Folder to output the results. If it does not exist it will be created.}

\item{configfile}{A file specifying the parameters of this function (without \code{inputfolder}, \code{outputfolder} and \code{configfile}). Having the parameters in a file can be handy if many samples with the same parameter settings are to be run. If a \code{configfile} is specified, it will take priority over the command line parameters.}

\item{numCPU}{The numbers of CPUs that are used. Should not be more than available on your machine.}

\item{reuse.existing.files}{A logical indicating whether or not existing files in \code{outputfolder} should be reused.}

\item{windowsize}{The window size used to calculate deltaWs, either an integer or 'scale'.}

\item{pairedEndReads}{Set to \code{TRUE} if you have paired-end reads in your file.}

\item{chromosomes}{If only a subset of the chromosomes should be processed, specify them here.}

\item{min.mapq}{Minimum mapping quality when importing from BAM files.}

\item{trim}{The amount of outliers in deltaWs removed to calculate the stdev (10 will remove top 10\% and bottom 10\% of deltaWs).}

\item{peakTh}{The treshold that the peak deltaWs must pass to be considered a breakpoint.}

\item{zlim}{The number of stdev that the deltaW must pass the peakTh (ensures only significantly higher peaks are considered).}

\item{bg}{The amount of background introduced into the genotype test.}

\item{minReads}{The minimum number of reads required for genotyping.}
}
\description{
This function is an easy-to-use wrapper to find breakpoints with \code{\link[breakpointR]{runBreakpointr}} in parallel.
}
\author{
Aaron Taudt
}

