#' Generates a bedfile from inFile
#'
#' Write a bedfile from Breakpoint.R files for upload on to UCSC Genome browser
#'
#' @param fileName is used to index the bedfile(s)
#' @param fileDestination is location to write bedfile(s)
#' @param inFile A \code{\link{GRanges}} object with strand and mapq metadata, such as that generated by \code{\link{bam2GRanges}}
#' @param deltaGraph If \code{TRUE}, will generate a bedgraph from a \code{\link{GRanges}} object with metadata column "deltaW" generated by \code{\link{deltaWCalculator}}
#' @param breakTrack If given a \code{\link{GRanges}} object with metadata "genoT" (e.g. newBreaks) will write a bedtrack with refined breakpoints
#' @param summary If \code{TRUE} appends these data to a file for each index (with all chromosomes included) 
#' @author Ashley Sanders, David Porubsky, Aaron Taudt
#' @export

writeBedFile <- function(fileName, dataDirectory, inFile, deltaGraph=F, breakTrack=NULL, bin='NA', summary=F) {
  message("Writing BED")  	
	
  fileDestination <- file.path(dataDirectory, "")
	
  index <- strsplit(fileName, '\\.')[[1]][1] # pulls out first element in fileName
  
  # write bedfile of reads that in deltaWs GRanges object 
  chr <- inFile$chromosome[1]	
  bedfile <- as.data.frame(inFile)[c('chromosome','start','end', 'mapq', 'strand')]
  bedfile$name <- 0 # adds a column of 0 as the'name' in the bedfile
  bedfile <- bedfile[c('chromosome','start','end','name','mapq', 'strand')]
  head <- paste('track name=', chr, '_reads_', index, ' visibility=1 colorByStrand="103,139,139 243,165,97"', sep="")
  write.table(head, file=paste(fileDestination, basename(fileName), '_', chr, '_bin', bin, '_breakpointR.bed', sep=""), row.names=F, col.names=F, quote=F, append=T, sep='\t')   
  write.table(bedfile, file=paste(fileDestination, basename(fileName), '_', chr, '_bin', bin, '_breakpointR.bed', sep=""), row.names=F, col.names=F, quote=F, append=T, sep='\t')
  
  if(deltaGraph == T){
    ## append the bedgraph of deltaWs:
    bedG <- as.data.frame(inFile)[c('chromosome','start','end','deltaW')]
    head <- paste('track type=bedGraph name=dWs', '_bin', bin, '_', chr, '_', index,' description=BedGraph_of_deltaWs_',basename(fileName), '_',  chr, ' visibility=full color=200,100,10', sep="")
    write.table(head, file=paste(fileDestination, basename(fileName), '_', chr, '_bin', bin, '_breakpointR.bed', sep=""), row.names=FALSE, col.names=F, quote=F, append=T, sep='\t')   
    write.table(bedG, file=paste(fileDestination, basename(fileName), '_', chr, '_bin', bin, '_breakpointR.bed', sep=""), row.names=FALSE, col.names=F, quote=F, append=T, sep='\t')      
  }

#  if(length(breakTrack) != 1) # if breakTracks contains a GRanges object 
#  { ## append the breakpoints to the file:
    if (is.null(breakTrack)) {	
      bpG<-data.frame(chr,'0','1','noBreaks')      		
    } else {	
      #breakTrack <- insertchr(breakTrack)
      bpG<-  as.data.frame(breakTrack)[c('chromosome','start','end','genoT')]
    }  
    head<- paste('track name=BPs_bin', bin, '_', chr, '_', index, ' description=BedGraph_of_breakpoints_',basename(fileName), '_',  chr, ' visibility=pack color=75,125,180', sep="")
    write.table(head, file=paste(fileDestination, basename(fileName), '_', chr, '_bin', bin, '_breakpointR.bed', sep=""), row.names=FALSE, col.names=F, quote=F, append=T, sep='\t')   
    write.table(bpG, file=paste(fileDestination, basename(fileName), '_', chr, '_bin', bin, '_breakpointR.bed', sep=""), row.names=FALSE, col.names=F, quote=F, append=T, sep='\t')
#  }
  
  if(summary == T){		
    SummaryfileDestination <- file.path(dataDirectory, 'Summary_files/')
	if (!file.exists(SummaryfileDestination)) {
		dir.create(SummaryfileDestination)
        }	
	
    ### Write an summary file of deltaWs and breaks for ALL CHR of the fileName
    if( chr== 'chr1'){
      head_reads<- paste('track name=', index, '_reads visibility=1 colorByStrand="103,139,139 243,165,97"', sep="")
      write.table(head_reads, file=paste(SummaryfileDestination, 'AllChr_', basename(fileName), '_reads.txt', sep=""), row.names=FALSE, col.names=F, quote=FALSE, append=F, sep='\t')   
      head_dW<- paste('track type=bedGraph name=', index,'_dWs description=BedGraph_of_deltaWs_',basename(fileName), '_allChr visibility=full color=200,100,10', sep="")
      write.table(head_dW, file=paste(SummaryfileDestination, 'AllChr_', basename(fileName), '_DeltaWs.txt', sep=""), row.names=FALSE, col.names=F, quote=FALSE, append=F, sep='\t')   
      head_br<- paste('track name=', index, '_BPs description=BedGraph_of_breakpoints_',basename(fileName),'_allChr visibility=dense color=75,125,180', sep="")
      write.table(head_br,  file=paste(SummaryfileDestination, 'AllChr_', basename(fileName), '_breakPoints.txt', sep=""), row.names=FALSE, col.names=F, quote=FALSE, append=F, sep='\t')
    } # write data
    write.table(bedfile,file=paste(SummaryfileDestination, 'AllChr_', basename(fileName), '_reads.txt', sep=""), row.names=FALSE, col.names=F, quote=FALSE, append=T, sep='\t')   
    write.table(bedG, file=paste(SummaryfileDestination, 'AllChr_', basename(fileName), '_DeltaWs.txt', sep=""), row.names=FALSE, col.names=F, quote=FALSE, append=T, sep='\t')
    write.table(bpG, file=paste(SummaryfileDestination, 'AllChr_', basename(fileName), '_breakPoints.txt', sep=""), row.names=FALSE, col.names=F, quote=FALSE, append=T, sep='\t')
  }
}
